kimchi/src/folding/error_term.rs

```diff
@@ -54,17 +55,27 @@ pub(crate) fn eval_sided<'a, C: FoldingConfig>(
         Add(e1, e2) => eval_exp_error(e1, env, side) + eval_exp_error(e2, env, side),
         Sub(e1, e2) => eval_exp_error(e1, env, side) - eval_exp_error(e2, env, side),
         Mul(e1, e2) => {
-            let d1 = e1.folding_degree();
-            let d2 = e2.folding_degree();
-            let e1 = match d1 {
-                Degree::Two => eval_sided(e1, env, side),
-                _ => eval_exp_error(e1, env, side),
-            };
-            let e2 = match d2 {
-                Degree::Two => eval_sided(e2, env, side),
-                _ => eval_exp_error(e2, env, side),
-            };
-            e1 * e2
+            //this assumes to some degree that selectors don't multiply each other
+            let selector = check_selector(e1)
+                .or(check_selector(e2))
+                .zip(env.enabled_selector())
+                .map(|(s1, s2)| s1 == s2);
+            match selector {
+                Some(false) => EvalLeaf::Result(env.inner.zero_vec()),
```

Comment:
indeed, in some way predicts the evaluation of subtrees and avoids computing them, at some point I thought of a more general approach that ignores any expressions that will end up in zero, not just involving selectors.
I would likely even be simpler to use, but at that point it was not worth the time, it may be worth doing in any future refactor.